### PR TITLE
BackPropagation test with one hidden layer.

### PR DESCRIPTION
**Pull request by cashecollins**
Back propagation test with one hidden layer and various training inputs, I wasn't sure how to get our information nor was I sure about the structure of our network so I just have 1 hidden layer, but we can modify it if we want to have more.

### CODE REVIEW

**Comment by ktrinanm** [comment]
I would remove the .class file from the commit.. i also suggest making sure this class can be easily integrated with the other code we have so far.

**Comment by cashecollins** [comment]
So I know you are all asleep at this point, so we can talk about it tomorrow, but I have a couple questions about the structure/number of hidden layers that we are using. As well as how we are accessing those nodes.

**Comment by ktrinanm** [comment]
Hey I just looked at this, but I'm going to wait until you add any changes you made today before I review it. Thanks!

**Comment by ktrinanm** [comment]
Hey where are you at on finishing the integration changes?

File: backpropagation.java [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+        double y[] = new double[numClasses]; //actual array is given by outside method for random weights to start.
+
+        // Unit errors.
+        double erro[] = new double[numClasses]; // erro array is the array that stores all of the changes that need to be made to the weights attached to an output node
+        double errh[] = new double[hiddenNeurons]; // errh array is the array that stores the changes that need to be made between hidden layers; again this will need to be changed depending on how we handle hidden layers.
+
+        int[] r = new int[2]; //Target is the expected output neuron for each data set.
+        z = neurons[1];
+
+        if (classification[n] == 1){
```
Comment (ktrinanm) [inline backpropagation.java]: So, here, you're reading from the classification array and not from the arrayList that has the training data. That could be causing problems.

File: backpropagation.java [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+    }
+
+    private static double sigmoidDerivative(final double val){
+        return (val * (1.0 - val));
+    }
+
+
+    //Linear Algebra Helper Functions
+    private static double vectorTransposeMultiplication(double[] A, double[] B){
+        double result = 0.0;
```
Comment (ktrinanm) [inline backpropagation.java]: Before completing this, I would verify that A.length == B.length because it could bring an issue to light if that's false.

**Review by ktrinanm — COMMENTED** [review]
Maybe checkout these comments to see if the mentioned lines could be causing issues?

File: backpropagation.java (hunk 1 above)
Comment (cashecollins) [reply backpropagation.java]: Thanks! I made the change.

File: backpropagation.java [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+        //layer is the current layer, weather it be the input layer of weights or the hidden layer of weights
+        int layer = 0;//for now layer will be 0, or the input layer, because i'm not sure how our layers are working.
+        double[][] w = weights[layer+1]; //wih is weights between input and hidden layers.
+        double[][] v = weights[(weights.length-1)]; //who is weights of hidden to output
+        double deltaW[][] = new double[numInputs][hiddenNeurons];
+        double deltaV[][] = new double[hiddenNeurons][numClasses];
+
+        double sum = 0.0;
+        for (int h = 0; h < INPUT_NEURONS; h++){
+            double s = vectorTransposeMultiplication(weights[1][h], x);
```
[the remark this answers is not in the document]
Comment (cashecollins) [reply backpropagation.java]: nah, the way that our weights are initialized  wweights[0] is an array full of 0.0, weights[1] are the weights from input to hidden, and weights[2] are the weights from hidden to output.

[the remark this answers is not in the document]
Comment (ktrinanm) [reply backpropagation.java]: Interesting. That might be something we need to look into updating

File: backpropagation.java (hunk 1 above)
Comment (ktrinanm) [reply backpropagation.java]: Cool. Let me know if it helps. did talking to tadayon help at all?

File: backpropagation.java (hunk 2 above)
Comment (cashecollins) [reply backpropagation.java]: Alright I looked at this and made some changes so that the weights update to the correct spots.

File: backpropagation.java (hunk 1 above)
Comment (cashecollins) [reply backpropagation.java]: I was really busy and the free time that I had, he was busy so I wasn't able to talk to him

Comment (cashecollins) [reply backpropagation.java]: I've been debugging though and I think that the weights are updating correctly. except that like joe was saying i think we have doubles overflowing and rounding to 1.0 or 0.0 and that might be messing with the code, I tried changing the sigmoid scale and that helped a little but still there was some rounding.

Comment (ktrinanm) [reply backpropagation.java]: So how close are we to being able to get results from it?
